
2-1-1 
函数已经结束，线程依旧访问局部变量
```c++
struct func                       // 函数对象
{
  int& i;
  func(int& i_) : i(i_) {}
  void operator() ()              // 重载 () 操作符
  {
    for (unsigned j=0 ; j<1000000 ; ++j)
    {
      do_something(i);            // 1. 潜在访问隐患：悬空引用
    }
  }
};

void oops()
{
  int some_local_state=0;
  func my_func(some_local_state);
  std::thread my_thread(my_func);
  my_thread.detach();          // 2. 不等待线程结束 -> my_thread.join()
}                              // 3. 新线程可能还在运行
```
2-1-2 join(), joinable()

2-1-3
如果想要分离一个线程，可以在线程启动后，直接使用detach()进行分离。如果打算等待对应线程，则需要细心挑选调用join()的位置。当在线程运行之后产生异常，在join()调用之前抛出，就意味着这次调用会被跳过。
避免应用被抛出的异常所终止，就需要作出一个决定。通常，当倾向于在无异常的情况下使用join()时，需要在异常处理过程中调用join()，从而避免生命周期的问题。下面的程序清单是一个例子。
```c++
struct func; // 定义在清单2.1中
void f()
{
  int some_local_state=0;
  func my_func(some_local_state);
  std::thread t(my_func);
  try
  {
    do_something_in_current_thread();
  }
  catch(...)
  {
    t.join();  // 1
    throw;
  }
  t.join();  // 2
}
```


